### PR TITLE
create a "provisional cache" to restore performance in the case of cycles

### LINKED ISSUE #60846: forego caching cycles leads to a severe perf regression
#60444 has introduced a severe perf regression when building/testing the `conch-runtime` crate.

Previously a test run would take ~5 mins, and with the latest nightly (`rustc 1.36.0-nightly (372be4f36 2019-05-14)`) it now takes ~82(!!) mins.

<pre>
$ git clone https://github.com/ipetkov/conch-runtime.git
$ cd conch-runtime
$ cargo test --lib
   # snip
    Finished dev [unoptimized + debuginfo] target(s) in 4m 16s
$ cargo clean
$ cargo +nightly test --lib
   # snip
    Finished dev [unoptimized + debuginfo] target(s) in 82m 54s
</pre>

<details>
<summary>
Crate info
</summary>
The crate offers the functionality to execute shell programs. Each piece of the grammar is represented as a node which can hold generic sub-nodes. The reasoning for this is so that the crate consumer could customize their AST with different/custom nodes, while reusing existing implementations.

The shell grammar is deeply recursive. Basically each command can vary in complexity (compound commands such as `case`, `for`, or simple commands like `echo foo`), but is ultimately made up of a list of shell words (literals, interpolations, etc.). Because each word can contain a command substitution, the AST type is recursive (a `Command<W>` has a `Word<C>` type, which gives us `Command<Word<Command<...>>`).

There are two "top-level" type definitions which seek to unify the entire AST tree concretely which are basically `TopLevelCommand(Command<TopLevelWord>)` `TopLevelWord(Word<TopLevelCommand>)`.

The crate also heavily uses generics and trait bounds (perhaps overly so), however, there's hopefully some low hanging fruits that can reduce the 16x slow down in performance.
</details>

cc @nikomatsakis @pnkfelix

### PR DESCRIPTION
**Pull request by nikomatsakis**
Introduce a "provisional cache" that caches the results of auto trait resolutions but keeps them from entering the *main* cache until everything is ready. This turned out a bit more complex than I hoped, but I don't see another short term fix -- happy to take suggestions! In the meantime, it's very clear we need to rework the trait solver. This resolves the extreme performance slowdown experienced in #60846 -- I plan to add a perf.rust-lang.org regression test to track this.

Caveat: I've not run `x.py test` in full yet.

r? @pnkfelix 
cc @arielb1 

Fixes #60846

### CODE REVIEW

File: src/libsyntax_pos/lib.rs [hunk 1]
```diff
@@ -227,14 +227,6 @@ impl SpanData {
     }
 }
 
-// The interner is pointed to by a thread local value which is only set on the main thread
-// with parallelization is disabled. So we don't allow `Span` to transfer between threads
-// to avoid panics and other errors, even though it would be memory safe to do so.
-#[cfg(not(parallel_compiler))]
-impl !Send for Span {}
-#[cfg(not(parallel_compiler))]
-impl !Sync for Span {}
```
Comment (bjorn3) [inline src/libsyntax_pos/lib.rs]: Why are those removed? This wasnt a hack, but it ensured that `Span` is accessed from another thread, which would have a different span interner (or none). It is a logic bug to access a `Span` with a different span interner.

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/libsyntax_pos/lib.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/libsyntax_pos/lib.rs]: Er, yes, that was a mistake, thanks.

**Comment by nikomatsakis** [comment]
So I think I found the bug causing cycle-cache-err-60010.rs to unexpectedly pass -- copy-and-paste error. Working on a fix.

**Comment by nikomatsakis** [comment]
I think the problem is fixed, not sure what's up with travis

File: src/librustc/traits/select.rs [hunk 2]
```diff
@@ -3934,28 +4003,283 @@ impl<'o, 'tcx> TraitObligationStack<'o, 'tcx> {
...
+///
+/// During evaluation, we consult this provisional cache and rely on
+/// it. Accessing a cached value is considered equivalent to accessing
+/// a result at `reached_depth`, so it marks the *current* solution as
+/// provisional as well. If an error is encountered, we toss out any
+/// provisional results added from the subtree that encountered the
+/// error.  When we pop the node at `reached_depth` from the stack, we
+/// can commit all the things that remain in the provisional cache.
+struct ProvisionalEvaluationCache<'tcx> {
+    /// next "depth first number" to issue -- just a counter
```
Comment (pnkfelix) [inline src/librustc/traits/select.rs]: nit: add `(DFN)` after `"depth first number" as a way of informally binding the acronym in a manner that makes it show up when one does a (case-sensitive) search to find such a binding in the docs.

Comment (pnkfelix) [reply src/librustc/traits/select.rs]: (oh I guess it does show up in this "binding" fashion below in the doc for `struct ProvisionalEvaluation`

Comment (pnkfelix) [reply src/librustc/traits/select.rs]: You might also explicitly state that it is "depth-first number" (and not just "depth") is that this identifies where a node falls in the graph when laid out in depth-first order, right? (I briefly thought "there must be a standard term for this notion that is less awkward, but I didn't see anything better than "DFS number", which does not seem superior to "DFN")

File: src/libsyntax_pos/lib.rs (hunk 1 above)
Comment (pnkfelix) [reply src/libsyntax_pos/lib.rs]: (I assume I should not `r+` until these are put back...)

**Comment by pnkfelix** [comment]
none of my feedback is severe enough to block an r+, so I'll go ahead and mark it so.

**Comment by mati865** [comment]
`mingw-check` runs with [`RUN_CHECK_WITH_PARALLEL_QUERIES=1`](https://github.com/rust-lang/rust/blob/96636f3eb8206f7499451ebc5b347d000d7aee08/src/ci/docker/mingw-check/Dockerfile#L22) and seems to hang like it did before https://github.com/rust-lang/rust/pull/60967/commits/3ed05613eeadd6baf332f6d4021c1973cb37ac21

**Comment by Zoxc** [comment]
You'll need `cfg(bootstrap)` on the workaround in https://github.com/rust-lang/rust/commit/3ed05613eeadd6baf332f6d4021c1973cb37ac21.

File: src/librustc/traits/select.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/librustc/traits/select.rs]: DFN is terminology that I took from various prolog papers -- probably this one:

> Efficient Top-Down Computation of Queries Under the Well-formed Semantics (Chen, Swift, and Warren; Journal of Logic Programming '95)

so it's not *totally* unstandard =)

**Comment by nikomatsakis** [comment]
@mati865 @Zoxc I confess I didn't actually *try* building a parallel compiler -- did one of you try it successfully?

File: src/librustc/traits/select.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/librustc/traits/select.rs]: Actually, I'm not sure why i'm tracking *both* depth *and* dfn -- I think you only need the DFN.

Comment (nikomatsakis) [reply src/librustc/traits/select.rs]: I think I just started with depth and had to add DFN later and didn't realize I could simplify it.

**Comment by mati865** [comment]
Unfortunately I'm out of time to play with it this week.

**Comment by Zoxc** [comment]
Building a parallel compiler and documenting it works with is PR. So I can do try builds again after this lands.

**Comment by howaboutsynergy** [comment]
<details>
<summary>

Compiled parallel rustc with this `config.toml`(click me to expand):

</summary>

```toml
[llvm]
optimize = true
release-debuginfo = false 
assertions = false 
ccache = "/usr/bin/ccache" 
static-libstdcpp = false
ninja = true
targets = "X86"
experimental-targets = ""
link-jobs = 0
[build]
build = "x86_64-unknown-linux-gnu"
host = ["x86_64-unknown-linux-gnu"]
target = ["x86_64-unknown-linux-gnu"] 
docs = true
compiler-docs = false 
submodules = true 
fast-submodules = true
locked-deps = false
full-bootstrap = false
extended = true 
tools = [ "src", "clippy", "rustfmt", "rls" ]
verbose = 2
sanitizers = false
profiler = false
low-priority = true
local-rebuild = false
print-step-timings = true
[install]
prefix = "/home/user/build/2nonpkgs/rust.stuff/rust/rust.installed.dir"
bindir = "bin"
[rust]
optimize = true
debug = false 
codegen-units = 1 
codegen-units-std = 1
debug-assertions = false 
debuginfo-level = 2
debuginfo-level-rustc = 2
debuginfo-level-std = 2
debuginfo-level-tools = 2
backtrace = true
incremental = true 
parallel-compiler = true 
channel = "dev"
rpath = true
verbose-tests = false
optimize-tests = true
codegen-tests = false 
ignore-git = false 
lld = false
deny-warnings = true 
backtrace-on-ice = true
llvm-libunwind = false 
[target.x86_64-unknown-linux-gnu]
[dist]
src-tarball = false
missing-tools = false
```

</details>

That is, rust `commit e699ea096fcc2fc9ce8e8bcf884e11496a31cc9f (HEAD -> master, origin/master, origin/HEAD)` + https://github.com/rust-lang/rust/pull/61754.patch

```
Build completed successfully in 0:28:13

real	28m14.051s
user	112m4.553s
sys	1m42.036s

// and with ./x.py install altogether:
Compile(failed?:0) took 3202 secs aka 53 minutes 22 seconds. date='Fri 14 Jun 2019 03:58:19 PM CEST'
//fabricate is a single-threaded slow-poke!! not-a-freakin'-fan
```

Using steps from https://github.com/rust-lang/rust/issues/60846#issue-444283743 with former rust:
```
/tmp/conch-runtime 
$ cargo clean; time cargo test --lib
...
^C  Building [======================================================> ] 99/100: conch-runtime                                 
real	7m6.081s
user	1m41.996s
sys	0m5.601s

//yes I C-c-ed it!

$ rustv
!! former-master-installed (default)
!! Executing '/home/user/.cargo/bin/rustc' in pwd='/home/user' with args(1): '-vV'
rustc 1.37.0-dev (cdd743755 2019-06-13)
binary: rustc
commit-hash: cdd743755adef8eb27410a9f4e480d3cf74abeaf
commit-date: 2019-06-13
host: x86_64-unknown-linux-gnu
release: 1.37.0-dev
LLVM version: 8.0
!! former-master-installed (default)
!! Executing '/home/user/build/2nonpkgs/rust.stuff/cargo/cargo//target/release//cargo' in pwd='/home/user' with args(1): '-vV'
cargo 1.37.0-dev (fa05862c 2019-06-13)
release: 1.37.0
commit-hash: fa05862cd0c6b899b801fda0f256ac5b9bae69d9
commit-date: 2019-06-13
```

and with patched rust:
```
$ cargo clean; time cargo test --lib
...
    Finished dev [unoptimized + debuginfo] target(s) in 31.62s
     Running target/debug/deps/conch_runtime-403b4e0e8e40cfbe

running 24 tests
test env::args::tests::test_args ... ok
test env::args::tests::test_name ... ok
test env::args::tests::test_set_args ... ok
test env::args::tests::test_shift_args ... ok
test env::args::tests::test_sub_env_no_needless_clone ... ok
test env::fd::tests::test_sub_env_no_needless_clone ... ok
test env::fd::tests::test_set_get_and_close_file_desc ... ok
test env::fd::tests::test_set_and_closefile_desc_in_child_env_should_not_affect_parent ... ok
test env::func::tests::test_set_and_unset_function_in_child_should_not_affect_parent ... ok
test env::func::tests::test_set_function_in_parent_visible_in_child ... ok
test env::func::tests::test_set_get_unset_function ... ok
test env::func::tests::test_sub_env_no_needless_clone ... ok
test env::last_status::tests::test_set_last_status_in_child_env_should_not_affect_parent ... ok
test env::last_status::tests::test_env_set_and_get_last_status ... ok
test env::var::tests::test_env_vars ... ok
test env::var::tests::test_set_get_unset_exported_var ... ok
test env::var::tests::test_get_env_vars_visible_in_parent_and_child ... ok
test env::var::tests::test_sub_env_no_needless_clone ... ok
test env::var::tests::test_set_var_in_child_env_should_not_affect_parent ... ok
test env::var::tests::test_set_get_unset_var ... ok
test error::tests::ensure_runtime_errors_are_send_and_sync ... ok
test io::tests::ensure_file_desc_is_send_and_sync ... ok
test env::env_impl::tests::test_env_is_interactive ... ok
test io::pipe::tests::smoke ... ok

test result: ok. 24 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out


real	0m31.642s
user	2m14.046s
sys	0m6.977s

//oh my gawd! wha?!! somebody pinch me

$ rustv
!! master-installed (default)
!! Executing '/home/user/.cargo/bin/rustc' in pwd='/home/user' with args(1): '-vV'
rustc 1.37.0-dev (e699ea096 2019-06-14)
binary: rustc
commit-hash: e699ea096fcc2fc9ce8e8bcf884e11496a31cc9f
commit-date: 2019-06-14
host: x86_64-unknown-linux-gnu
release: 1.37.0-dev
LLVM version: 8.0
!! master-installed (default)
!! Executing '/home/user/build/2nonpkgs/rust.stuff/cargo/cargo//target/release//cargo' in pwd='/home/user' with args(1): '-vV'
cargo 1.37.0-dev (fa05862c 2019-06-13)
release: 1.37.0
commit-hash: fa05862cd0c6b899b801fda0f256ac5b9bae69d9
commit-date: 2019-06-13
```

**Comment by howaboutsynergy** [comment]
That fail\* has been happening since at least [10 days ago](https://github.com/rust-lang/rust/pull/59331#issuecomment-499294942).

\* this fail:
`[01:38:57] test process::tests::uid_works ... test process::tests::uid_works has been running for over 60 seconds`
`No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.`

```rust
      #[test]
      #[cfg_attr(target_os = "android", ignore)]
      #[cfg(unix)]
      fn uid_works() {
          use crate::os::unix::prelude::*;
  
          let mut p = Command::new("/bin/sh")
                              .arg("-c").arg("true")
                              .uid(unsafe { libc::getuid() })
                              .gid(unsafe { libc::getgid() })
                              .spawn().unwrap();                                                                                
          assert!(p.wait().unwrap().success());
      }
```
file: `src/libstd/process.rs`

Now why would that happen?

**Comment by Zoxc** [comment]
Is that a flaky `xsv` test @rust-lang/infra ?

**Comment by Mark-Simulacrum** [comment]
It doesn't look like it should be based on the amount of panics, but I'm not sure. It does use quickcheck which introduces a level of randomness which could lead to it being spurious and/or not related to this PR in particular (cc @BurntSushi as the author).

Since I suspect it is unrelated to this PR, let's @bors retry xsv spurious test

**Comment by pnkfelix** [comment]
Given the number of beta-regressions that we believe were injected by PR #60444 (see #61960), I am beta-nominating this PR.

However, I am also a little nervous about back-porting this PR to the current beta, which (if I understand correctly) is scheduled to become the stable release in about two weeks, but his PR itself only landed on nightly 3 days ago. So: Do not immediately assume my nomination is also an endorsement for a backport; I am just trying to initiate discussion of the matter.

**Comment by pnkfelix** [comment]
discussed at T-compiler meeting. Given that this just landed in nightly, we're going to give this a week to bake before officially accepting for beta backport. (But our current inclination is to accept.)

**Comment by Mark-Simulacrum** [comment]
@rust-lang/compiler We need to get beta backports approved a little earlier this cycle as we're promoting beta->stable on Saturday. If we could get an FCP going for beta-accepted on this PR that'd be great; or if someone wants to unilaterally beta accept that works too.

**Comment by pnkfelix** [comment]
I'm going to unilaterally beta-accept this.

We [did talk about the beta-nom of #61754](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/weekly.20meeting.202019-06-20.20.2354818/near/168602206) at the T-compiler meeting last week, and the main misgivings we had were: 1. it didn't seem to resolve *all* of the performance and overflow issues, 2. it was a bit big+risky for a backport, and 3. it hadn't had much time to bake.

But, now it has had time to bake, and (as also stated at the meeting), it fixes a critical issue.

So, accepting for beta-backport.

**Comment by pnkfelix** [comment]
This should have had the beta-nominated flag removed as part of PR #62065, right?
